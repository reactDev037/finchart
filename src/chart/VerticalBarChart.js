import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { scaleLinear, scaleBand } from 'd3-scale';
import { max } from 'd3-array';

import ChartBase from './ChartBase';
import XAxis from '../axis/XAxis';
import YAxis from '../axis/YAxis';
import VerticalBarGroup from '../shape/VerticalBarGroup';

/**
 * component:
 *   rotated ordinal XAxis without labels in the tick
 *   horizontal YAxis
 *   XAxis label on top of the chart
 *   BarGroup with Label rotated
 */

/**
 * Construct a BarChart from a catogorised data set
 * Using Ordinal Scale(scaleBand) to show catogories on xAxis
 */
export class Component extends PureComponent {
  static propTypes = {
    className: PropTypes.string,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    padding: PropTypes.number,
    headroom: PropTypes.number,
    barGroupPadding: PropTypes.number,
    data: PropTypes.arrayOf(PropTypes.object).isRequired,
    fields: PropTypes.shape({
      x: PropTypes.string.isRequired,
      y: PropTypes.string.isRequired,
    }).isRequired,
    axisNames: PropTypes.shape({
      x: PropTypes.string,
      y: PropTypes.string,
    }),
  };

  static defaultProps = {
    barGroupPadding: 30,
    headroom: 0.2,
    padding: 60,
    axisNames: {
      x: null,
      y: null,
    },
  };

  render() {
    const {
      className,
      width,
      height,
      padding,
      headroom,
      barGroupPadding,
      data,
      fields,
      axisNames,
    } = this.props;
    const xDataAbs = data.map(d => d[fields.x]).map(d => Math.abs(d));
    const xDomainAbs = max(xDataAbs) * (1 + headroom);
    const xDomain = [-xDomainAbs, xDomainAbs];
    const yData = data.map(d => d[fields.y]);
    const xScale = scaleLinear().domain(xDomain).range([0, width]);
    const yScale = scaleBand()
      .domain(yData)
      .range([0, height])
      .padding(barGroupPadding);
    const transform = `translate(${xScale(0)}, 0)`;
    return (
      <ChartBase {...{ className, width, height, padding }}>
        <text className="yAxisName" dy={-8} {...{ transform }}>
          {axisNames.y}
        </text>
        <VerticalBarGroup {...{ width, xScale, yScale, data, fields }} />
        <XAxis {...{ height, width, xScale }} axisName={axisNames.x} />
        <YAxis {...{ height, width, yScale, transform }} notick />
      </ChartBase>
    );
  }
}

export default styled(Component)`
  .yAxisName {
    fill: black;
    font-weight: bold;
    text-anchor: middle;
    font-size: 11px;
  }
`;
