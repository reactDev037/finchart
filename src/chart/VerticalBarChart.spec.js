import React from 'react';

import VerticalBarChart from './VerticalBarChart';

describe('<VerticalBarChart />', () => {
  it('renders', () => {
    const chartConfig = {
      width: 640,
      height: 220,
      data: [
        { region: 'Nordic Countries', volume: 30 },
        { region: 'Eastern Europe', volume: -20 },
        { region: 'Latin America', volume: 40 },
        { region: 'Asia', volume: -20 },
        { region: 'Middle East', volume: -30 },
        { region: 'Western Europe', volume: 40 },
      ],
      fields: {
        x: 'volume',
        y: 'region',
      },
      axisNames: {
        x: 'Any X Name',
        y: 'Any Y Name',
      },
    };
    const wrapper = render(<VerticalBarChart {...chartConfig} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles data input format error', () => {
    const chartConfig = {
      width: 960,
      height: 320,
      data: [
        { region: 'CN', volume: 30 },
        { region: 'BE', volume: 20 },
        { region: 'TX', volume: 40 },
        { region: 'JD', volume: 20 },
        { region: 'DF', volume: 10 },
        { region: 'OD', volume: 40 },
      ],
      fields: {
        x: 'type',
        y: 'volume',
      },
    };
    const wrapper = render(<VerticalBarChart {...chartConfig} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles missing data properly', () => {
    // TODO: using tickArguments
  });

  it('handles partial data key pollution', () => {
    // TODO: using tickArguments
  });
});
