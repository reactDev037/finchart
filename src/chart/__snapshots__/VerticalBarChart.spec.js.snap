// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerticalBarChart /> handles data input format error 1`] = `
<svg
  class="sc-ifAKCX buPUHa sc-bdVaJa jgDyUg"
  height="320"
  width="960"
>
  <rect
    class="background"
    height="320"
    width="960"
  />
  <text
    class="yAxisName"
    dy="-8"
    transform="translate(NaN, 0)"
  />
  <text>
    Input Data Format Error
  </text>
  <g
    class="xaxis-sc-bwzfXH fjjIqd"
    fill="none"
    font-family="sans-serif"
    font-size="10"
    text-anchor="middle"
    transform="translate(0,320)"
  >
    <text
      class="axisName"
      transform="translate(480, 30)"
    />
    <path
      class="domain"
      d="M0.5,0V0.5H960.5V0"
      stroke="#000"
    />
  </g>
  <g
    class="yaxis-sc-htpNat jpKklr"
    fill="none"
    font-family="sans-serif"
    font-size="10"
    text-anchor="end"
    transform="translate(NaN, 0)"
  >
    <text
      class="axisName"
      transform="translate(-25, 160) rotate(270)"
    />
    <path
      class="domain"
      d="M0,0.5H0.5V320.5H0"
      stroke="#000"
    />
    <g
      class="tick"
      opacity="1"
      transform="translate(0,64.5)"
    >
      <line
        stroke="#000"
        x2="0"
      />
      <text
        dy="0.32em"
        fill="#000"
        x="-5"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,128.5)"
    >
      <line
        stroke="#000"
        x2="0"
      />
      <text
        dy="0.32em"
        fill="#000"
        x="-5"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,192.5)"
    >
      <line
        stroke="#000"
        x2="0"
      />
      <text
        dy="0.32em"
        fill="#000"
        x="-5"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,256.5)"
    >
      <line
        stroke="#000"
        x2="0"
      />
      <text
        dy="0.32em"
        fill="#000"
        x="-5"
      />
    </g>
  </g>
</svg>
`;

exports[`<VerticalBarChart /> renders 1`] = `
<svg
  class="sc-ifAKCX buPUHa sc-bdVaJa jgDyUg"
  height="220"
  width="640"
>
  <rect
    class="background"
    height="220"
    width="640"
  />
  <text
    class="yAxisName"
    dy="-8"
    transform="translate(320, 0)"
  >
    Any Y Name
  </text>
  <g
    class="bargroup-sc-bxivhb djSFBb"
    transform="translate(0, -10)"
  >
    <rect
      height="20"
      width="200"
      x="320"
      y="31.42857142857143"
    />
    <rect
      class="negative"
      height="20"
      width="133.33333333333331"
      x="186.66666666666669"
      y="62.85714285714286"
    />
    <rect
      height="20"
      width="266.66666666666663"
      x="320"
      y="94.28571428571428"
    />
    <rect
      class="negative"
      height="20"
      width="133.33333333333331"
      x="186.66666666666669"
      y="125.71428571428571"
    />
    <rect
      class="negative"
      height="20"
      width="200"
      x="120"
      y="157.14285714285714"
    />
    <rect
      height="20"
      width="266.66666666666663"
      x="320"
      y="188.57142857142856"
    />
    <text
      alignment-baseline="middle"
      dx="-5"
      dy="11"
      x="320"
      y="31.42857142857143"
    >
      Nordic Countries
    </text>
    <text
      alignment-baseline="middle"
      class="negative"
      dx="5"
      dy="11"
      x="320"
      y="62.85714285714286"
    >
      Eastern Europe
    </text>
    <text
      alignment-baseline="middle"
      dx="-5"
      dy="11"
      x="320"
      y="94.28571428571428"
    >
      Latin America
    </text>
    <text
      alignment-baseline="middle"
      class="negative"
      dx="5"
      dy="11"
      x="320"
      y="125.71428571428571"
    >
      Asia
    </text>
    <text
      alignment-baseline="middle"
      class="negative"
      dx="5"
      dy="11"
      x="320"
      y="157.14285714285714"
    >
      Middle East
    </text>
    <text
      alignment-baseline="middle"
      dx="-5"
      dy="11"
      x="320"
      y="188.57142857142856"
    >
      Western Europe
    </text>
  </g>
  <g
    class="xaxis-sc-bwzfXH fjjIqd"
    fill="none"
    font-family="sans-serif"
    font-size="10"
    text-anchor="middle"
    transform="translate(0,220)"
  >
    <text
      class="axisName"
      transform="translate(320, 30)"
    >
      Any X Name
    </text>
    <path
      class="domain"
      d="M0.5,0V0.5H640.5V0"
      stroke="#000"
    />
    <g
      class="tick"
      opacity="1"
      transform="translate(53.83333333333333,0)"
    >
      <line
        stroke="#000"
        y2="0"
      />
      <text
        dy="0.71em"
        fill="#000"
        y="8"
      >
        -40
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(120.5,0)"
    >
      <line
        stroke="#000"
        y2="0"
      />
      <text
        dy="0.71em"
        fill="#000"
        y="8"
      >
        -30
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(187.16666666666669,0)"
    >
      <line
        stroke="#000"
        y2="0"
      />
      <text
        dy="0.71em"
        fill="#000"
        y="8"
      >
        -20
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(253.83333333333331,0)"
    >
      <line
        stroke="#000"
        y2="0"
      />
      <text
        dy="0.71em"
        fill="#000"
        y="8"
      >
        -10
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(320.5,0)"
    >
      <line
        stroke="#000"
        y2="0"
      />
      <text
        dy="0.71em"
        fill="#000"
        y="8"
      >
        0
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(387.16666666666663,0)"
    >
      <line
        stroke="#000"
        y2="0"
      />
      <text
        dy="0.71em"
        fill="#000"
        y="8"
      >
        10
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(453.83333333333337,0)"
    >
      <line
        stroke="#000"
        y2="0"
      />
      <text
        dy="0.71em"
        fill="#000"
        y="8"
      >
        20
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(520.5,0)"
    >
      <line
        stroke="#000"
        y2="0"
      />
      <text
        dy="0.71em"
        fill="#000"
        y="8"
      >
        30
      </text>
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(587.1666666666666,0)"
    >
      <line
        stroke="#000"
        y2="0"
      />
      <text
        dy="0.71em"
        fill="#000"
        y="8"
      >
        40
      </text>
    </g>
  </g>
  <g
    class="yaxis-sc-htpNat jpKklr"
    fill="none"
    font-family="sans-serif"
    font-size="10"
    text-anchor="end"
    transform="translate(320, 0)"
  >
    <text
      class="axisName"
      transform="translate(-25, 110) rotate(270)"
    />
    <path
      class="domain"
      d="M0,0.5H0.5V220.5H0"
      stroke="#000"
    />
    <g
      class="tick"
      opacity="1"
      transform="translate(0,31.92857142857143)"
    >
      <line
        stroke="#000"
        x2="0"
      />
      <text
        dy="0.32em"
        fill="#000"
        x="-5"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,63.35714285714286)"
    >
      <line
        stroke="#000"
        x2="0"
      />
      <text
        dy="0.32em"
        fill="#000"
        x="-5"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,94.78571428571428)"
    >
      <line
        stroke="#000"
        x2="0"
      />
      <text
        dy="0.32em"
        fill="#000"
        x="-5"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,126.21428571428571)"
    >
      <line
        stroke="#000"
        x2="0"
      />
      <text
        dy="0.32em"
        fill="#000"
        x="-5"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,157.64285714285714)"
    >
      <line
        stroke="#000"
        x2="0"
      />
      <text
        dy="0.32em"
        fill="#000"
        x="-5"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(0,189.07142857142856)"
    >
      <line
        stroke="#000"
        x2="0"
      />
      <text
        dy="0.32em"
        fill="#000"
        x="-5"
      />
    </g>
  </g>
</svg>
`;
